### DISCUSSION TITLE
New must-have features

### DISCUSSION THREAD

**Discussion by wessleyn** [vkbo/novelWriter#2052] [Ideas]
Disclaimer: I did use Ai to simplify my explanations, cause am not a native of the English language . So I didn't troll you,😃

Hi, I’d like to suggest some improvements to streamline the editing process in NovelWriter. These small enhancements can go a long way in making the software more efficient for writers:

Suggested Features:

1. Move Lines Up and Down

Shortcut: (Ctrl+Shift+Arrow in VSCode)

This allows users to quickly reorganize their text by moving entire lines up or down without the need to cut and paste.

2. Delete an Entire Line

Shortcut: (Alt+D in VSCode)

A command to delete an entire line at once, simplifying the process of cleaning up unwanted lines.

3. Copy Lines Up or Down

Shortcut: (Ctrl+Shift+D in VSCode)

A command to duplicate the current line or selected lines, moving them either above or below the original line.

4. Auto Formatter for Line Spacing and Indentation

Automatically format text to ensure consistent spacing between lines and at the start of paragraphs. This will help maintain a clean structure, reducing manual corrections for spacing issues.

---

Potential Features (If Not Already Available):

5. Search All Instances of a Tag

Ability to search and display all scenes, chapters, or volumes where a particular tag (e.g., @volume, @milestone, @char) is used. This would be incredibly useful for tracking characters, themes, or plot points across the entire novel.

6. Custom Tagging System per Document

Enable users to create custom tags within each document for better organization and referencing. This would complement the existing structure and allow for more flexibility in categorizing content.

---

New Feature Suggestion: Visualization of Plot Spread

7. Narrative Plot Spread Visualization (Milestones/Plots/Scenes)

This idea might seem a bit ambitious, but hear me out. It would be helpful to visualize how plot points, milestones, or themes are distributed across the novel, using a bar chart or similar graphical representation.

Example Use Case:
Writers could track to what extent or portion of the novel a specific plot or theme is spread. For instance, you could visualize how much of the narrative is covered by each milestone or scene across chapters or volumes, much like a bar diagram. This could help in balancing the pacing and distribution of key story elements.

![1729233933491.jpg](https://github.com/user-attachments/assets/e9342e45-6fe5-4289-82bb-614851e572cf)

(Example of how plot spread could be visualized in a bar diagram)

How it Might Work:
Each scene could have tags that indicate which chapters or volumes it spans. By using this data, a visual bar graph could be generated, showing the distribution of milestones, premises, or ideas. This could be calculated based on the number of chapters or scenes covered by each element.

Further Potential:
A more complex graph could even show the setup, conflict, and resolution phases for each major plot arc, providing a high-level overview of the story structure.

**Comment by vkbo** [comment]
Again, please don't tag GitHub users in your posts.

> 1. Move Lines Up and Down

See #2007.

> 2. Delete an Entire Line
> 3. Copy Lines Up or Down

Duplicating lines isn't really a text writing feature, but more a programming feature, and is currently not planned. Deleting an entire line may be a useful feature.

> 4. Auto Formatter for Line Spacing and Indentation

I don't understand this one. Paragraphs have one empty line between them, and there is no way for the editor to know when you intend to start a new paragraph. Indentation is not supported, so I don't get that one either. Unless you mean the `>` and `<` indentation markers and, maybe it would be better to have a select + format feature for doing this on a range of paragraphs.

> 5. Search All Instances of a Tag

If by tag here you mean reference keyword, like `@plot`, `@location`, etc, these are already listed in the references panel under the document viewer window. See: https://novelwriter.io/docs/usage_writing.html#viewing-a-document

This panel also lists all documents pointing back to the current open document's `@tag` definition.

> 6. Custom Tagging System per Document

This will likely not be implemented unless the entire tags and reference system is redesigned from scratch. Since the tags are free form, you can use all sorts of prefixes and namespacing for the tags to categorise them. There will be more tools to navigate tags and references though, but their implementation has not been decided on yet.

> 7. Narrative Plot Spread Visualization (Milestones/Plots/Scenes)

Story structure tracking has been discussed, and a design plan decided upon. This will be implemented. The feature request is #1784, and the discussion of the implementation in #1769.

**Reply by wessleyn** [reply]
> 4. Auto Formatter for Line Spacing and Indentation
 
While typing, I noticed that you only allowed one-spacing  and the rest is labeled redundant ( hitting space more than once) as to the documentation (this should be red underline.  If while typing the editor automatically removes that space, it think this is what they call auto formatting.

Also when you leave space at the end of a line - green wavy line. I know this isn't allowed too .. while I can go line by line deleting or do the find and replace . Where I select the redundant space and replace with nothing. If the editor is smart enough to delete for me . This is what am referring to as auto formatting

Let's say I paste text into the editor, if the  app  heavy lift all the space readjustments then it'll be a great help

**Reply by vkbo** [reply]
The green underline is an optional feature to show where you have redundant spaces. It isn't an "error", hence it isn't red.

Auto-formatting can in principle fix double-spaces as you type, but that definitely has to be opt-in as many people actually use that in the English speaking world. Apparently it was common on type writers, and people insist on still using it.

Yes, you can search/replace `"  "` (two spaces) with `" "` one space, and fix it. I'm considering adding a text cleanup tool that can fix many small issues, like spaces and obviously missing capitalisation, etc. See #1045

**Reply by wessleyn** [reply]
> 6. Custom Tagging System per Document
 
Here what I really meant was, this tags are root level meaning if I have a dir structure like

plot -- milestones - milestone1


I can't tag milestones one because it's deeply nested and I have to move it into the root folder of plot and yet it retains the broad category @plot

@plot: milestone1

Here I can't even use the specific milestone name since I wouldn't know what type of note it is 

This might be more definitive  under @milestone 

Where I can say @milestione: getsBetrayed

If as you say the whole codebase might need reworking then can you at least add tags for second level dir structures


plot - milestones - mile1


@milestones: mile1, ...

This would still be visible in the reference viewer

**Reply by vkbo** [reply]
I repeat: Please don't tag random GitHub users in your posts. You keep doing that, and I've pointed it out several times now.

> I can't tag milestones one because it's deeply nested and I have to move it into the root folder of plot and yet it retains the broad category `@plot`
> 
> `@plot: milestone1`
> 
> Here I can't even use the specific milestone name since I wouldn't know what type of note it is

That's why I'm saying people are using the tags and references system for things it was never intended for. It is not designed for nested categories, and there is no clean way of implementing it. All you can do is include your subcategories in the tag definition.

> If as you say the whole codebase might need reworking then can you at least add tags for second level dir structures

That is precisely the thing the system does not support, so there is no "at least" here without rewriting the system from scratch. As I said, it was never intended for this type of usage.

**Reply by wessleyn** [reply]
> I repeat: Please don't tag random GitHub users in your posts. You keep doing that, and I've pointed it out several times now.
> 
> > I can't tag milestones one because it's deeply nested and I have to move it into the root folder of plot and yet it retains the broad category `@plot`
> > 
> > `@plot: milestone1`
> > 
> > Here I can't even use the specific milestone name since I wouldn't know what type of note it is
> 
> That's why I'm saying people are using the tags and references system for things it was never intended for. It is not designed for nested categories, and there is no clean way of implementing it. All you can do is include your subcategories in the tag definition.
> 
> > If as you say the whole codebase might need reworking then can you at least add tags for second level dir structures
> 
> That is precisely the thing the system does not support, so there is no "at least" here without rewriting the system from scratch. As I said, it was never intended for this type of usage.
> 
> 

It's really not intended cause I use GitHub mobile app and they're two ways to reply quote reply and a general level reply...am using quote reply which I believe is a specific reply on a post unless otherwise. Please advice

**Reply by vkbo** [reply]
Either add a slash after the `@` like this: @\plot, or wrap them in backticks to make them a code block.

**Reply by vkbo** [reply]
Anyway, it is not the reply/quote feature that is the problem, but that you type the @ + name/keyword, which nearly always corresponds to actual GitHub users.

**Reply by wessleyn** [reply]
@vkbo you mean like this

**Reply by vkbo** [reply]
Yes, don't do that.

**Reply by wessleyn** [reply]
> I repeat: Please don't tag random GitHub users in your posts. You keep doing that, and I've pointed it out several times now.
> 
> > I can't tag milestones one because it's deeply nested and I have to move it into the root folder of plot and yet it retains the broad category `@plot`
> > 
> > `@plot: milestone1`
> > 
> > Here I can't even use the specific milestone name since I wouldn't know what type of note it is
> 
> That's why I'm saying people are using the tags and references system for things it was never intended for. It is not designed for nested categories, and there is no clean way of implementing it. All you can do is include your subcategories in the tag definition.
> 
> > If as you say the whole codebase might need reworking then can you at least add tags for second level dir structures
> 
> That is precisely the thing the system does not support, so there is no "at least" here without rewriting the system from scratch. As I said, it was never intended for this type of usage.
> 
> 

I think I figured an alternative 


If my dir structure is  : 

Plot - milestones -  DescentIntoADevil (Against the Gods - yun che ref)


Then while I can't say "@/milestones: DescentIntoADevil"

When I reference this milestone in a volume let's say , " All thoughts to the devil volume in against the gods- by mars gravity"


This wouldn't be possible for reasons already mentioned... But

I can write this in the "DescentIntoADevil " milestones


# Descent Into A Devil 

@/tag: Milestone -- Descent Into A Devil 


And in the dir  (Novel - All thoughts to the devil --- Chapters) volume1 assuming you transformed it into a project note 


# Volume 7(yes, i didn't check): All thoughts to the devil 


@/plot: Milestone -- Descent Into A Devil
                   ^

Where the milestone here is a subdir and it actually works putting atag in any folder however deeply nested it might be and you can ref to it using the root level tag

As shown below

![descent.png](https://github.com/user-attachments/assets/c59a0c20-1a0e-4072-a28d-1b0f8f2bbc46)

**Reply by vkbo** [reply]
Yes, that's the proposed solution.

By the way, there is no point specifying both `%Short` and `%Synopsis`. They are aliases for the same thing. Short is meant for notes, and Synopsis for novel text.
